kong/runloop/balancer/balancers.lua

```diff
@@ -0,0 +1,549 @@
+      return balancers_M.create_balancer(upstream), upstream
+    end
+  end
+
+  return balancer, upstream
+end
+
+
+function balancers_M.create_balancers()
+  local all_upstreams, err = upstreams.get_all_upstreams()
+  if not all_upstreams then
+    log(CRIT, "failed loading initial list of upstreams: ", err)
+    return
+  end
+
+  local oks, errs = 0, 0
+  for ws_and_name, id in pairs(all_upstreams) do
+    local name = sub(ws_and_name, (find(ws_and_name, ":", 1, true)))
+
+    local upstream = upstreams.get_upstream_by_id(id)
+    local ok
+    if upstream ~= nil then
+      ok, err = balancers_M.create_balancer(upstream)
+    end
+    if ok ~= nil then
+      oks = oks + 1
+    else
+      log(CRIT, "failed creating balancer for ", name, ": ", err)
+      errs = errs + 1
+    end
+  end
+  log(DEBUG, "initialized ", oks, " balancer(s), ", errs, " error(s)")
+end
+
+
+--------- balancer object methods
+
+function balancer_mt:eachAddress(f, ...)
+  for _, target in ipairs(self.targets) do
+    for _, address in ipairs(target.addresses) do
+      f(address, target, ...)
+    end
+  end
+end
+
+function balancer_mt:findAddress(ip, port, hostname)
+  for _, target in ipairs(self.targets) do
+    if target.name == hostname then
+      for _, address in ipairs(target.addresses) do
+        if address.ip == ip and address.port == port then
+          return address
+        end
+      end
+    end
+  end
+end
+
+
+function balancer_mt:setAddressStatus(address, available)
+  if type(address) ~= "table"
+    or type(address.target) ~= "table"
+    or address.target.balancer ~= self
+  then
+    return nil, "not a known address"
+  end
+
+  if address.available == available then
+    return true, "already set"
+  end
+
+  address.available = available
+  local delta = address.weight
+  if available then
+    delta = -delta
+  end
+  address.target.unavailableWeight = address.target.unavailableWeight + delta
+  self.unavailableWeight = self.unavailableWeight + delta
+  self:updateStatus()
+  return true
+end
+
+
+function balancer_mt:disableAddress(target, entry)
+  -- from host:disableAddress()
+  local address = self:changeWeight(target, entry, 0)
+  if address then
+    address.disabled = true
+  end
+end
+
+
+local function setHostHeader(addr)
+  local target = addr.target
+
+  if target.nameType ~= "name" then
+    -- hostname is an IP address
+    addr.hostHeader = nil
+  else
+    -- hostname is an actual name
+    if addr.ipType ~= "name" then
+      -- the address is an ip, so use the hostname as header value
+      addr.hostHeader = target.name
+    else
+      -- the address itself is a nested name (SRV)
+      if addr.useSRVname then
+        addr.hostHeader = addr.ip
+      else
+        addr.hostHeader = target.name
+      end
+    end
+  end
+end
+
+function balancer_mt:addAddress(target, entry)
+  -- from host:addAddress
+  if type(entry) ~= "table"
+    or type(target) ~= "table"
+    or target.balancer ~= self
+  then
+    return
```

Comment:
Should this throw an error?

Reply:
yes, it would be better.

of course, it wouldn't do any difference, since the calling code (at least from target/update_dns_result) doesn't check the return.  but that's a good candidate for refactoring, I'd like the update_dns_result function to be much shorter and easier to read, with most of the duplicates/updating complexities moved into some high-level balancer/target methods.